cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,456 @@
+			fwCache:      newFakeFirewallCacheEmpty(),
+			inboundRules: []godo.InboundRule{fakeInboundRule},
+			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(fakeWorkerFWName, fakeInboundRule), newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)}, newFakeOKResponse(), nil
+			},
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					listFunc:     test.expectedGodoFirewallListResp,
+					addRulesFunc: test.expectedGodoFirewallAddRulesResp,
+					updateFunc:   test.expectedGodoFirewallUpdateResp,
+					createFunc:   test.expectedGodoFirewallCreateResp,
+					getFunc:      test.expectedGodoFirewallGetResp,
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			err := fc.fwManager.Set(ctx, test.inboundRules)
+			if want, got := test.expectedError, err; want != got {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+		})
+	}
+}
+
+func TestFirewallController_createInboundRules(t *testing.T) {
+	inboundRule := godo.InboundRule{
+		Protocol:  "tcp",
+		PortRange: "31220",
+	}
+	nodePortService := &v1.Service{
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeNodePort,
+			Ports: []v1.ServicePort{
+				{
+					Protocol: "tcp",
+					Port:     31220,
+					NodePort: 31220,
+				},
+			},
+		},
+	}
+	fakeServiceList := []*v1.Service{nodePortService}
+
+	testcases := []struct {
+		name        string
+		firewall    *godo.Firewall
+		fwCache     firewallCache
+		serviceList []*v1.Service
+	}{
+		{
+			name:        "successfully updates port range",
+			firewall:    nil,
+			fwCache:     newFakeFirewallCache(inboundRule),
+			serviceList: fakeServiceList,
+		},
+		{
+			name:        "return node port inbound rules when we find a service of type nodeport",
+			firewall:    newFakeFirewall(fakeWorkerFWName, inboundRule),
+			fwCache:     newFakeFirewallCache(inboundRule),
+			serviceList: fakeServiceList,
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			// setup
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{},
+			)
+			inboundRules := []godo.InboundRule{inboundRule}
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			rules := fc.createInboundRules(test.serviceList)
+			if !cmp.Equal(rules[0].PortRange, inboundRules[0].PortRange) {
+				t.Errorf("got %q, want %q", rules[0].PortRange, inboundRules[0].PortRange)
```

Comment:
This error message is identical to the one in line 409 except for the placeholder values. We should include what kind of value failed the assertion (i.e., port range here and protocol below).

That said, if we start validating that multiple inbound rules are processed correctly, then I think we should be able to use `cmp.Diff` against a slice of inbound rules directly.